lib/ts/utils/dateProvider/defaultImplementation.ts

```diff
@@ -0,0 +1,59 @@
+/* Copyright (c) 2024, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import WindowHandlerReference from "../windowHandler";
+
+export class DateProvider {
+    private static instance?: DateProvider;
+    public static readonly CLOCK_SKEW_KEY = "__st_clockSkewInMillis";
+    private clockSkewInMillis: number = 0;
+
+    // The static init method is used to create a singleton instance of DateProvider,
+    // as we require access to localStorage for initializing clockSkewInMillis.
+    // Access to localStorage is available only after WindowHandlerReference is initialized.
+    static init() {
+        if (DateProvider.instance !== undefined) {
+            return;
+        }
+
+        DateProvider.instance = new DateProvider();
+        const localStorage = WindowHandlerReference.getReferenceOrThrow().windowHandler.localStorage;
+        const stored = localStorage.getItemSync(DateProvider.CLOCK_SKEW_KEY);
+        const clockSkewInMillis = stored !== null ? parseInt(stored, 10) : 0;
+        DateProvider.instance.setClientClockSkewInMillis(clockSkewInMillis);
+    }
+
+    static getReferenceOrThrow(): DateProvider {
+        if (DateProvider.instance === undefined) {
+            throw new Error("DateProvider must be initialized before calling this method.");
+        }
+
+        return DateProvider.instance;
+    }
+
+    setClientClockSkewInMillis(clockSkewInMillis: number): void {
+        this.clockSkewInMillis = clockSkewInMillis;
+        const localStorage = WindowHandlerReference.getReferenceOrThrow().windowHandler.localStorage;
```

Comment:
i don't think we should use localstorage cause of people sharing session across sub domains. So we should:
- use frontend cookies; OR
- ideally, no need to store this value in a storage, and just store in memory. If it's not there in memory, can't we calculate it based on the access token payload and then save it in memory?

Reply:
Upon several discussions, we decided to keep the caching strategy while documenting all the considered cases.